### ISSUE TITLE
No way to catch opencv exception from managed code using Linux

### ISSUE THREAD

**Issue by AStueben** [shimat/opencvsharp#517]
# Summary of your issue
Under Linux, calling a opencvsharp function which then throws an exception on the native side causes the calling thread to crash. Apparently, this problem arises when passing an exception over the p/invoke boundary, see
https://github.com/dotnet/coreclr/issues/9288 and https://github.com/dotnet/coreclr/issues/9288

It seems that one must catch the exception on the native side and find a solution to inform the managed code that an exception has happened.

# Environment

Ubuntu 18.04LTS
OpenCv 3.4.1

# What did you do when you faced the problem?

I tried to register an CvErrorCallback to the opncv redirectError under Linux to intercept the exception. This does not work, as opencv seems to throws an exception with or without a callback function.

Also, i tried to wrap every call to opencv in the native code of opencvsharp in a try-catch block and use a custom callback-function to get informed on the csharp side of the call if an exception had occurred. This only works with functions which returns void, so is not a solution.

## Example code:

```
using System;
using OpenCvSharp;

class Program
{
    static void Main()
    {
        Mat src = Cv2.ImRead("lenna.png", ImreadModes.GrayScale);
        Mat dst = new Mat();
        try
        {
            // opencv throws an exception when blocksize is even
            int blockSize = 2;
            Cv2.AdaptiveThreshold(src,dst,  255, OpenCvSharp.AdaptiveThresholdTypes.MeanC,
                OpenCvSharp.ThresholdTypes.Binary, blockSize, 3);
        }
        catch(Exception ex)
        {
            Console.WriteLine("Exception catched.");
        }
        Console.WriteLine("Finished.");
    }
}
```

## Output:

```
OpenCV(3.4.1) Error: Assertion failed (blockSize % 2 == 1 && blockSize > 
1) in adaptiveThreshold, file 
/home/stueben/Programs/opencv/modules/imgproc/src/thresh.cpp, line 1525
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(3.4.1) 
/home/stueben/Programs/opencv/modules/imgproc/src/thresh.cpp:1525: 
error: (-215) blockSize % 2 == 1 && blockSize > 1 in function 
adaptiveThreshold
The program '[14599] test_prog.dll' has exited with code 0 (0x0).
```

## What did you intend to be?

```
Exception catched.
Finished.
```

I am highly invested to find a solution for this problem under Linux, of course I want to find one that will ultimately contribute to this project and be accepted. 
Any ideas?

**Comment by andreaskoepf** [comment]
The 'trick' that is used by OpenCvSharp on Windows does unfortunately not work on Linux with .net Core, e.g. [ErrorHandlerThrowException ](https://github.com/shimat/opencvsharp/blob/5134fb2ab442a3579142cf03d73fb6c7123df9cb/src/OpenCvSharp/PInvoke/NativeMethods.cs#L200) together with [redirectError](https://github.com/shimat/opencvsharp/blob/f53838e9c26983f572113be3cf639c4312bebfd0/src/OpenCvSharpExtern/core.h#L73).

Other OpenCV wrappers also have to deal with this problem, e.g. the 'official' CV2 wrapper for Python wraps every call with a try-catch block (via macro), e.g. see [cv2.cpp#L92](https://github.com/opencv/opencv/blob/45dd575ed2c8458181cd0dd286fa22d1061a4065/modules/python/src2/cv2.cpp#L92). But a major difference between CV2 wrapper an the OpenCVSharp functions is that the Python-interface is naturally much more homogeneous, e.g. I think there is no way to do this with minor C macro-magic for OpenCVSharp...


### Sketch of idea for solution for OpenCVSharp (via auto-code generation)

The meta-data  of the existing static `[DllImport]` functions could be used to generate an additional C# and C++ a exception-handling layer.

This exception safe layer would consist of the following parts and require only minor changes in the existing OpenCVSharp Wrapper code:

1. (C wrapper C-Lib): Automatically generate a new 'exception safe wrapper function' for each C-export  that calls the already existing C wrapper-function inside a try catch block and catches exceptions. These wrapper functions could have a homogeneous signature with e.g. `bool` or status `int` return type and optionally (if non-void) a pointer to store the return value. The return value would indicate whether the function completed successfully (e.g. whether an exception was caught).
2. (C# side): Auto-Generate replacements for the existing `[DllImport]` methods which have exactly the same name and signature: These new static methods would call the exception-safe  C wrappers and interpret their status result and either hand out the return value or throw a .Net exception.
3. (C# side) Add a new static class that provides the `[DllImport]` signatures for the exception safe wrapper C functions. 

Advantage of the approach: The main parts of OpenCvSharp are not touched, e.g. all existing  C wrappers just stay the same and also on the C# side the result is a static class which looks exactly as the `NativeMethods` ... all tedious work is done by a code-generator that can in the same way be used in the future when new methods are added to OpenCV.

### Exception translation:
The actual exception information could for example be transported back to the C# side via the OpenCV error callback which also works on Linux (unfortunately not offering a way to cancel the C++ exception generation). The exception information could be stored in a thread-local storage that is read by the auto-generated exception C# static methods.
Another option would be to call call a new custom callback from within the C++ catch block or add an additional argument to each of the exception-safe methods that allows to store the exception information.

I personally would favor to pass the exception information via callback since this would avoid any string allocation problems, e.g. generic exception information could consist of two strings: 1. type of exception 2. exception message. The `error-callback` would simply provide these strings as `const char*` an the callback handler could simple copy them...

@shimat We could develop that solution and create a PR. Very important to me would be to get your opinion and potentially approval for such an approach. E.g. do you see any other feasible (maybe better) way to handle the exceptions with .Net Core (on Linux)? Probably it could be done in platform dependent way with some kind of general call-thunk / to execute the actual C wrapper with a registered try-catch handler .. but I think the auto-code generation approach is simpler and platform independent. Of course we would end up with a little bit bigger lib...

**Comment by shimat** [comment]
Thank you for your detailed information and suggestion. I approve your auto-code generation approach :+1: I have never heard about the problem. 

My concern is that the size of the library (OpenCvSharpExtern.dll/so/dylib) may be larger. Although I think it is still acceptable, I saw some opinions that the file size of the DLL is too big and can not be embedded in application. Perhaps I should separate the extern DLL by module.

**Comment by turowicz** [comment]
+1 from me

**Comment by turowicz** [comment]
When that's done, could you merge it to the v3.x tag?

**Comment by irug** [comment]
+1 from me

**Comment by turowicz** [comment]
Any updates on this @shimat ?

**Comment by shimat** [comment]
I recognize that this issue has been resolved by #758 (#523).

**Comment by turowicz** [comment]
@shimat thanks
